README.md
# slurm-debian

To clone the submodules:
 * `git submodule update --init --recursive`

To update the submodules:
 * `git submodule foreach --recursive git pull`

Build the debian package:
 * `cd slurm`
 * `git checkout slurm-18.08`
 * `ln -s ../debian .`
 * `dch -n` (if you want to increase the version number)
 * `debian/rules build`
 * `fakeroot debian/rules binary`


debian/changelog
slurm-llnl (18.08.3-3) UNRELEASED; urgency=medium

  * Update Slurm to 18.08.3.
  * Added dependency for libssh2-1-dev, x11 forwarding
  * remove configure options: --enable-front-end, --enable-multiple-slurmd

 -- Bas van der Vlies <bas@test3.irc.sara.nl.irc.sara.nl>  Tue, 30 Oct 2018 15:42:47 +0100

slurm-llnl (18.08.1-1) stable; urgency=medium

  * Update Slurm to 18.08.1.

 -- Martijn Kruiten <martijn.kruiten@surfsara.nl>  Wed, 10 Oct 2018 13:52:09 +0200

slurm-llnl (17.11.9-1) UNRELEASED; urgency=medium

  * Update Slurm to 17.11.9
  * Build libpam_slurm_adopt.so

 -- Martijn Kruiten <martijn.kruiten@surfsara.nl>  Mon, 03 Sep 2018 15:35:31 +0200

slurm-llnl (17.11.7-4) UNRELEASED; urgency=medium

  * /var/spool/slurm-llnl must owned by slurm.slurm 

 -- Bas van der Vlies <bas.vandervlies@surfsara.nl>  Tue, 17 Jul 2018 16:04:05 +0200

slurm-llnl (17.11.7-3) UNRELEASED; urgency=medium

  * Change uid/gid to 555. No clash with CUA
  * some small update to the stable 17.11 branch
  * create /var/spool/slurm-llnl in client package

 -- Bas van der Vlies <bas.vandervlies@surfsara.nl>  Tue, 17 Jul 2018 14:48:02 +0200

slurm-llnl (17.11.7-2) UNRELEASED; urgency=medium

  * Updated debian package info this version replace all version <= 17.02

 -- Bas van der Vlies <bas@test3.irc.sara.nl.irc.sara.nl>  Thu, 21 Jun 2018 10:03:12 +0200

slurm-llnl (17.11.7-1) UNRELEASED; urgency=low

  * New 17.11.7 version

 -- Bas van der Vlies <bas.vandervlies@surfsara.nl>  Fri, 15 Jun 2018 12:05:15 +0200

slurm-llnl (17.02.9-1) unstable; urgency=medium

  * New upstream release fixes CVE-2017-15566 (Closes: #880530)
  * Add new symbols to libslurm31 and libslurmdb31
  * Modify header patch to remove google script in the footer
  * Rename header patch

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Sun, 05 Nov 2017 01:02:24 +0100

slurm-llnl (17.02.7-1) unstable; urgency=medium

  * New upstream release
  * Rightly moving slurmdbd symbols in slurmdbd debug package
  * Switch from -dbg packages to automatic -dbgsym
  * Change package priority to optional as required by the new standard
  * Change d/copyright format uri to https
  * Standard version upgraded to 4.1.1
  * Increase debhelper version dependency to get rid of dh-systemd
  * Add srun, sbatch and sacct autopkgtests

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Sat, 07 Oct 2017 01:25:22 +0200

slurm-llnl (17.02.5-1) unstable; urgency=medium

  [ Mehdi Dogguy and Gennaro Oliva ]
  * New upstream release.
  * Refresh patches.
  * Update libslurm{db,}31.symbols

  [ Gennaro Oliva ]
  * Update d/copyright and switch to copyright-format 1.0

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Mon, 03 Jul 2017 15:30:51 +0200

slurm-llnl (17.02.1.2-1) unstable; urgency=medium

  * New upstream major release
  * Update libslurm{,db} package files to soname 31
  * Refresh patches and remove those included upstream
  * Remove srun_cr man page not installed without blcr support
  * Update service files to avoid killing job (Closes: #855905)

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Wed, 08 Mar 2017 08:57:37 +0100

slurm-llnl (16.05.9-1) unstable; urgency=medium

  * New upstream release
  * Overrides spelling-error-in-binary false positives

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Fri, 03 Feb 2017 09:50:02 +0100

slurm-llnl (16.05.8-1) unstable; urgency=medium

  * New upstream release 
  * Remove wiki.conf.5 from slurm-client manpages

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Sat, 07 Jan 2017 02:40:23 +0100

slurm-llnl (16.05.7-2) unstable; urgency=medium

  * Enable multiple slurmd support (Closes: #849140)
  * Drop dependency on openssl-blacklist (Closes: #848627)
  * slurmd and slurm-client postrm don't kill slurm processes anymore
  * Checks whether the slurm user exists before deleting
  * Use numerical uid 64030 to kill slurm owned processes

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Fri, 30 Dec 2016 16:06:07 +0100

slurm-llnl (16.05.7-1) unstable; urgency=medium

  * New upstream release (Closes: #828549)
  * Add libswitch-perl dependency to slurm-wlm-torque (Closes: #847342)
  * Add slurmd.README.Debian (Closes: #845268) 
  * Move README.Debian to slurmctld.README.Debian
  * Add lintian overrides for no bindnow in many packages
  * Use jquery.min.js provided by libjs-jquery instead of upstream's
  * Add jquery source in missing-sources

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Wed, 14 Dec 2016 11:01:06 +0100

slurm-llnl (16.05.6-1) unstable; urgency=medium

  * Update libslurm{,db} symbols files
  * Remove unused hardening-no-bindnow lintian overrides
  * Add lsb-base dependency to slurmdbd
  * Change libmysqlclient-dev build dependency to default
  * Add libpmi2-0 packages thanks to Markus Geimer (Closes: #840404)

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Fri, 28 Oct 2016 16:23:07 +0200

slurm-llnl (16.05.5-1) unstable; urgency=medium

  * New upstream release
  * Update debian configurators
  * Update header patch to remove google script in html
  * Update symbol files

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Fri, 30 Sep 2016 16:08:43 +0200

slurm-llnl (16.05.4-1) unstable; urgency=medium

  * New upstream release 
  * Add missing sh5util binary to slurm-client and qalter qrerun and
    mpiexec.slurm to slurm-wlm-torque (as requested by Jordi Blasco)
  * Update ax_lib_hdf5.m4 to avoid CPPFLAGS to be overwritten
  * Add HDF5_CPPFLAGS to successfully compile libsh5util_old.la
  * Restore default fortify flags

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Mon, 26 Sep 2016 11:44:23 +0200

slurm-llnl (16.05.2-1) unstable; urgency=medium

  * New upstream release 

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Fri, 29 Jul 2016 18:34:26 +0200

slurm-llnl (16.05.0-1) unstable; urgency=medium

  * New upstream major release 
  * Update library files to version 30

 -- Gennaro Oliva <oliva.g@na.icar.cnr.it>  Tue, 14 Jun 2016 11:15:17 +0200

slurm-llnl (15.08.11-1) unstable; urgency=medium

  * New upstream release 

debian/slurmctld.init.d
#!/bin/sh
#
# chkconfig: 345 90 10
# description: SLURM is a simple resource management system which \
#              manages exclusive access o a set of compute \
#              resources and distributes work to those resources.
#
# processname: /usr/sbin/slurmctld
# pidfile: /var/run/slurm-llnl/slurmctld.pid
#
# config: /etc/default/slurmctld
#
### BEGIN INIT INFO
# Provides:          slurmctld
# Required-Start:    $remote_fs $syslog $network munge
# Required-Stop:     $remote_fs $syslog $network munge
# Should-Start:      $named
# Should-Stop:       $named
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: slurm daemon management
# Description:       Start slurm to provide resource management
### END INIT INFO

BINDIR=/usr/bin
CONFDIR=/etc/slurm-llnl
LIBDIR=/usr/lib
SBINDIR=/usr/sbin

# Source slurm specific configuration
if [ -f /etc/default/slurmctld ] ; then
    . /etc/default/slurmctld
else
    SLURMCTLD_OPTIONS=""
fi

# Checking for slurm.conf presence
if [ ! -f $CONFDIR/slurm.conf ] ; then
    if [ -n "$(echo $1 | grep start)" ] ; then
      echo Not starting slurmctld
    fi
      echo slurm.conf was not found in $CONFDIR
      echo Please follow the instructions in \
            /usr/share/doc/slurmctld/README.Debian
    exit 0
fi


DAEMONLIST="slurmctld"
test -f $SBINDIR/slurmctld || exit 0

#Checking for lsb init function
if [ -f /lib/lsb/init-functions ] ; then
  . /lib/lsb/init-functions
else
  echo Can\'t find lsb init functions
  exit 1
fi

# setup library paths for slurm and munge support
export LD_LIBRARY_PATH=$LIBDIR${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}

#Function to check for cert and key presence
checkcertkey()
{
  MISSING=""
  keyfile=""
  certfile=""

  if [ "$1" = "slurmctld" ] ; then
    keyfile=$(grep JobCredentialPrivateKey $CONFDIR/slurm.conf | grep -v "^ *#")
    keyfile=${keyfile##*=}
    keyfile=${keyfile%#*}
    [ -e $keyfile ] || MISSING="$keyfile"
  fi

  if [ "${MISSING}" != "" ] ; then
    echo Not starting slurmctld
    echo $MISSING not found
    echo Please follow the instructions in \
      /usr/share/doc/slurmctld/README.cryptotype-openssl
    exit 0
  fi
}

get_daemon_description()
{
    case $1 in
      slurmd)
        echo slurm compute node daemon
	;;
      slurmctld)
	echo slurm central management daemon
	;;
      *)
	echo slurm daemon
	;;
    esac
}

start() {
  CRYPTOTYPE=$(grep CryptoType $CONFDIR/slurm.conf | grep -v "^ *#")
  CRYPTOTYPE=${CRYPTOTYPE##*=}
  CRYPTOTYPE=${CRYPTOTYPE%#*}
  if [ "$CRYPTOTYPE" = "crypto/openssl" ] ; then
    checkcertkey $1
  fi

  # Create run-time variable data
  mkdir -p /var/run/slurm-llnl
  chown slurm:slurm /var/run/slurm-llnl

  # Checking if StateSaveLocation is under run
  if [ "$1" = "slurmctld" ] ; then
    SDIRLOCATION=$(grep StateSaveLocation /etc/slurm-llnl/slurm.conf \
                       | grep -v "^ *#")
    SDIRLOCATION=${SDIRLOCATION##*=}
    SDIRLOCATION=${SDIRLOCATION%#*}
    if [ "${SDIRLOCATION}" = "/var/run/slurm-llnl/slurmctld" ] ; then
      if ! [ -e /var/run/slurm-llnl/slurmctld ] ; then
        ln -s /var/lib/slurm-llnl/slurmctld /var/run/slurm-llnl/slurmctld
      fi
    fi
  fi

  desc="$(get_daemon_description $1)"
  log_daemon_msg "Starting $desc" "$1"
  unset HOME MAIL USER USERNAME
  #FIXME $STARTPROC $SBINDIR/$1 $2
  STARTERRORMSG="$(start-stop-daemon --start --oknodo \
                   --exec "$SBINDIR/$1" -- $2 2>&1)"
  STATUS=$?
  log_end_msg $STATUS
  if [ "$STARTERRORMSG" != "" ] ; then
    echo $STARTERRORMSG
  fi
  touch /var/lock/slurm
}

stop() {
    desc="$(get_daemon_description $1)"
    log_daemon_msg "Stopping $desc" "$1"
    STOPERRORMSG="$(start-stop-daemon --oknodo --stop -s TERM \
                    --exec "$SBINDIR/$1" 2>&1)"
    STATUS=$?
    log_end_msg $STATUS
    if [ "$STOPERRORMSG" != "" ] ; then
      echo $STOPERRORMSG
    fi
    rm -f /var/lock/slurm
}

getpidfile() {
    dpidfile=`grep -i ${1}pid $CONFDIR/slurm.conf | grep -v '^ *#'`
    if [ $? = 0 ]; then
        dpidfile=${dpidfile##*=}
        dpidfile=${dpidfile%#*}
    else
        dpidfile=/var/run/${1}.pid
    fi

    echo $dpidfile
}

#
# status() with slight modifications to take into account
# instantiations of job manager slurmd's, which should not be
# counted as "running"
#
slurmstatus() {
    base=${1##*/}

    pidfile=$(getpidfile $base)

    pid=`pidof -o $$ -o $$PPID -o %PPID -x $1 || \
         pidof -o $$ -o $$PPID -o %PPID -x ${base}`

    if [ -f $pidfile ]; then
        read rpid < $pidfile
        if [ "$rpid" != "" -a "$pid" != "" ]; then
            for i in $pid ; do
                if [ "$i" = "$rpid" ]; then
                    echo "${base} (pid $pid) is running..."
                    return 0
                fi
            done
        elif [ "$rpid" != "" -a "$pid" = "" ]; then
#           Due to change in user id, pid file may persist
#           after slurmctld terminates
            if [ "$base" != "slurmctld" ] ; then
               echo "${base} dead but pid file exists"
            fi
            return 1
        fi

    fi

    if [ "$base" = "slurmctld" -a "$pid" != "" ] ; then
        echo "${base} (pid $pid) is running..."
        return 0
    fi

    echo "${base} is stopped"

    return 3
}

#
# stop slurm daemons,
# wait for termination to complete (up to 10 seconds) before returning
#
slurmstop() {
    for prog in $DAEMONLIST ; do
       stop $prog
       for i in 1 2 3 4
       do
          sleep $i
          slurmstatus $prog
          if [ $? != 0 ]; then
             break
          fi
       done
    done
}

#
# The pathname substitution in daemon command assumes prefix and
# exec_prefix are same.  This is the default, unless the user requests
# otherwise.
#
# Any node can be a slurm controller and/or server.
#
case "$1" in
    start)
	start slurmctld "$SLURMCTLD_OPTIONS"
        ;;
    startclean)
        SLURMCTLD_OPTIONS="-c $SLURMCTLD_OPTIONS"
        start slurmctld "$SLURMCTLD_OPTIONS"
        ;;
    stop)
	slurmstop
        ;;
    status)
	for prog in $DAEMONLIST ; do
	   slurmstatus $prog
	done
        ;;
    restart)
        $0 stop
        $0 start
        ;;
    force-reload)
        $0 stop
        $0 start
	;;
    condrestart)
        if [ -f /var/lock/subsys/slurm ]; then
            for prog in $DAEMONLIST ; do
                 stop $prog
                 start $prog
            done
        fi
        ;;
    reconfig)
	for prog in $DAEMONLIST ; do
	    PIDFILE=$(getpidfile $prog)
	    start-stop-daemon --stop --signal HUP --pidfile \
	      "$PIDFILE" --quiet $prog
	done
	;;
    test)
	for prog in $DAEMONLIST ; do
	    echo "$prog runs here"
	done
	;;
    *)
        echo "Usage: $0 {start|startclean|stop|status|restart|reconfig|condrestart|test}"
        exit 1
        ;;
esac


debian/slurmd.init.d
#!/bin/sh
#
# chkconfig: 345 90 10
# description: SLURM is a simple resource management system which \
#              manages exclusive access o a set of compute \
#              resources and distributes work to those resources.
#
# processname: /usr/sbin/slurmd
# pidfile: /var/run/slurm-llnl/slurmd.pid
#
# config: /etc/default/slurmd
#
### BEGIN INIT INFO
# Provides:          slurmd
# Required-Start:    $remote_fs $syslog $network munge
# Required-Stop:     $remote_fs $syslog $network munge
# Should-Start:      $named
# Should-Stop:       $named
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: slurm daemon management
# Description:       Start slurm to provide resource management
### END INIT INFO

BINDIR=/usr/bin
CONFDIR=/etc/slurm-llnl
LIBDIR=/usr/lib
SBINDIR=/usr/sbin

# Source slurm specific configuration
if [ -f /etc/default/slurmd ] ; then
    . /etc/default/slurmd
else
    SLURMD_OPTIONS=""
fi

# Checking for slurm.conf presence
if [ ! -f $CONFDIR/slurm.conf ] ; then
    if [ -n "$(echo $1 | grep start)" ] ; then
      echo Not starting slurmd
    fi
      echo slurm.conf was not found in $CONFDIR
      echo Please follow the instructions in \
            /usr/share/doc/slurmd/README.Debian
    exit 0
fi


DAEMONLIST="slurmd"
test -f $SBINDIR/slurmd || exit 0

#Checking for lsb init function
if [ -f /lib/lsb/init-functions ] ; then
  . /lib/lsb/init-functions
else
  echo Can\'t find lsb init functions
  exit 1
fi

# setup library paths for slurm and munge support
export LD_LIBRARY_PATH=$LIBDIR${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}

#Function to check for cert and key presence
checkcertkey()
{
  MISSING=""
  keyfile=""
  certfile=""

  if [ "$1" = "slurmd" ] ; then
    keyfile=$(grep JobCredentialPublicCertificate $CONFDIR/slurm.conf \
                  | grep -v "^ *#")
    keyfile=${keyfile##*=}
    keyfile=${keyfile%#*}
    [ -e $keyfile ] || MISSING="$keyfile"
  fi

  if [ "${MISSING}" != "" ] ; then
    echo Not starting slurmd
    echo $MISSING not found
    echo Please follow the instructions in \
      /usr/share/doc/slurmd/README.cryptotype-openssl
    exit 0
  fi
}

get_daemon_description()
{
    case $1 in
      slurmd)
        echo slurm compute node daemon
	;;
      slurmctld)
	echo slurm central management daemon
	;;
      *)
	echo slurm daemon
	;;
    esac
}

start() {
  CRYPTOTYPE=$(grep CryptoType $CONFDIR/slurm.conf | grep -v "^ *#")
  CRYPTOTYPE=${CRYPTOTYPE##*=}
  CRYPTOTYPE=${CRYPTOTYPE%#*}
  if [ "$CRYPTOTYPE" = "crypto/openssl" ] ; then
    checkcertkey $1
  fi

  # Create run-time variable data
  mkdir -p /var/run/slurm-llnl
  chown slurm:slurm /var/run/slurm-llnl

  # Checking if SlurmdSpoolDir is under run
  if [ "$1" = "slurmd" ] ; then
    SDIRLOCATION=$(grep SlurmdSpoolDir /etc/slurm-llnl/slurm.conf \
                       | grep -v "^ *#")
    SDIRLOCATION=${SDIRLOCATION##*=}
    SDIRLOCATION=${SDIRLOCATION%#*}
    if [ "${SDIRLOCATION}" = "/var/run/slurm-llnl/slurmd" ] ; then
      if ! [ -e /var/run/slurm-llnl/slurmd ] ; then
        ln -s /var/lib/slurm-llnl/slurmd /var/run/slurm-llnl/slurmd
      fi
    fi
  fi

  desc="$(get_daemon_description $1)"
  log_daemon_msg "Starting $desc" "$1"
  unset HOME MAIL USER USERNAME
  #FIXME $STARTPROC $SBINDIR/$1 $2
  STARTERRORMSG="$(start-stop-daemon --start --oknodo \
                   --exec "$SBINDIR/$1" -- $2 2>&1)"
  STATUS=$?
  log_end_msg $STATUS
  if [ "$STARTERRORMSG" != "" ] ; then
    echo $STARTERRORMSG
  fi
  touch /var/lock/slurm
}

stop() {
    desc="$(get_daemon_description $1)"
    log_daemon_msg "Stopping $desc" "$1"
    STOPERRORMSG="$(start-stop-daemon --oknodo --stop -s TERM \
                    --exec "$SBINDIR/$1" 2>&1)"
    STATUS=$?
    log_end_msg $STATUS
    if [ "$STOPERRORMSG" != "" ] ; then
      echo $STOPERRORMSG
    fi
    rm -f /var/lock/slurm
}

getpidfile() {
    dpidfile=`grep -i ${1}pid $CONFDIR/slurm.conf | grep -v '^ *#'`
    if [ $? = 0 ]; then
        dpidfile=${dpidfile##*=}
        dpidfile=${dpidfile%#*}
    else
        dpidfile=/var/run/${1}.pid
    fi

    echo $dpidfile
}

#
# status() with slight modifications to take into account
# instantiations of job manager slurmd's, which should not be
# counted as "running"
#
slurmstatus() {
    base=${1##*/}

    pidfile=$(getpidfile $base)

    pid=`pidof -o $$ -o $$PPID -o %PPID -x $1 || \
         pidof -o $$ -o $$PPID -o %PPID -x ${base}`

    if [ -f $pidfile ]; then
        read rpid < $pidfile
        if [ "$rpid" != "" -a "$pid" != "" ]; then
            for i in $pid ; do
                if [ "$i" = "$rpid" ]; then
                    echo "${base} (pid $pid) is running..."
                    return 0
                fi
            done
        elif [ "$rpid" != "" -a "$pid" = "" ]; then
#           Due to change in user id, pid file may persist
#           after slurmctld terminates
            if [ "$base" != "slurmctld" ] ; then
               echo "${base} dead but pid file exists"
            fi
            return 1
        fi

    fi

    if [ "$base" = "slurmctld" -a "$pid" != "" ] ; then
        echo "${base} (pid $pid) is running..."
        return 0
    fi

    echo "${base} is stopped"

    return 3
}

#
# stop slurm daemons,
# wait for termination to complete (up to 10 seconds) before returning
#
slurmstop() {
    for prog in $DAEMONLIST ; do
       stop $prog
       for i in 1 2 3 4
       do
          sleep $i
          slurmstatus $prog
          if [ $? != 0 ]; then
             break
          fi
       done
    done
}

#
# The pathname substitution in daemon command assumes prefix and
# exec_prefix are same.  This is the default, unless the user requests
# otherwise.
#
# Any node can be a slurm controller and/or server.
#
case "$1" in
    start)
	start slurmd "$SLURMD_OPTIONS"
        ;;
    startclean)
        SLURMD_OPTIONS="-c $SLURMD_OPTIONS"
        start slurmd "$SLURMD_OPTIONS"
        ;;
    stop)
	slurmstop
        ;;
    status)
	for prog in $DAEMONLIST ; do
	   slurmstatus $prog
	done
        ;;
    restart)
        $0 stop
        $0 start
        ;;
    force-reload)
        $0 stop
        $0 start
	;;
    condrestart)
        if [ -f /var/lock/subsys/slurm ]; then
            for prog in $DAEMONLIST ; do
                 stop $prog
                 start $prog
            done
        fi
        ;;
    reconfig)
	for prog in $DAEMONLIST ; do
	    PIDFILE=$(getpidfile $prog)
	    start-stop-daemon --stop --signal HUP --pidfile \
	      "$PIDFILE" --quiet $prog
	done
	;;
    test)
	for prog in $DAEMONLIST ; do
	    echo "$prog runs here"
	done
	;;
    *)
        echo "Usage: $0 {start|startclean|stop|status|restart|reconfig|condrestart|test}"
        exit 1
        ;;
esac
